doc/OnlineDocs/contributed_packages/edi/blackboxconstraints.rst

```diff
@@ -0,0 +1,473 @@
+
+In other words, if you wish to code a black-box constraint using EDI, you will be using the Runtime Constraint constructor.
+
+In this context, a *Black-Box* is defined as a routine that performs hidden computation not visible EDI, pyomo, or more generally the optimization algorithm.  However, it is **not** assumed that black-boxes are unable to return gradient information.  A black-box in this context may be capable of returning arbitrary derivative information.
+
+
+Construction
+------------
+
+Runtime constraints consist of two separate elements that need to be constructed.  
+
+
+Constructing a Black Box
+++++++++++++++++++++++++
+
+First, we need to create an object which is visible to pyomo/EDI that calls the black-box function.  EDI calls this a ``BlackBoxFunctionModel``, and it is a base class that gets inherited into the objects you will create as a user.
+
+A simple example is shown below:
+
+.. code-block:: python
+
+        import pyomo.environ as pyo
+        from pyomo.environ import units
+        from pyomo.contrib.edi import BlackBoxFunctionModel
+
+        class Parabola(BlackBoxFunctionModel):
+            def __init__(self): 
+
+                # Call parent init
+                super().__init__()
+
+                # A brief description of the model
+                self.description = 'This model evaluates the function: y = x**2'
+
+                # Append the model inputs
+                self.inputs.append(
+                    name='x', units='ft' , description='The x variable'
+                )
+
+                # Append the model outputs
+                self.outputs.append(
+                    name='y', units='ft**2' , description='The y variable'
+                )
+                
+                # Set the highest available derivative
+                # Should be 1 for most cases but defaults to 0
+                self.availableDerivative = 1
+
+            def BlackBox(self, x): # The actual function that does things
+
+                # Convert to correct units and cast to a float
+                x = pyo.value(units.convert(x,self.inputs['x'].units))
+
+                # Compute y
+                y = x**2
+
+                # Compute dy/dx
+                dydx = 2*x
+                
+                # Add the units to the output
+                y = y * self.outputs['y'].units
+                
+                # Add the units to the derivative for output
+                dydx = dydx * self.outputs['y'].units / self.inputs['x'].units
+                
+                # Return
+                return y, [dydx] # return z, grad(z), hess(z)...
```

Comment:
This is documented in BlackBox Constraints->Construction->The Black Box Method

Reply:
I could do smarter detection in my wrapper function if the black box is scalar input and scalar output, but this will cause frustrating singularities.  Will improve documentation to make this more clear

Reply:
Added a more explicit comment in the new commit